### ISSUE TITLE
Stop supporting Callable when resolving values

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#9778]
When resolving values we unpack different kinds of values:

* `Callable` and Groovy's `Closure` gets called
* `Provider` gets unwrapped
* etc.

~~We should also support `java.util.function.Supplier` in these places.~~

We should stop supporting `Callable` and require the use of `Provider` instead.

**Comment by wolfs** [comment]
I wonder if using `Supplier` in those cases really makes sense. We unpack `Closure` and the Kotlin lambdas, since it allows for a nicer DSL. `Provider`s are the native value holders, so we need to unpack them.

I don't see `Supplier` providing a better DSL for Java that what we already support with `Callable`, given that you are going to use a Lambda expression anyway.

Is there a particular use-cases you came across where using `Supplier`s would be useful?

**Comment by lptr** [comment]
The points you make make sense. But does this lead to the conclusion that Gradle APIs will always use `Callable` instead of `Supplier`, or is there a migration path?

**Comment by wolfs** [comment]
The Gradle API should mostly use `Action` and `Provider`, which both are Gradle types. It also uses `Transformer` for function types. The [factory methods for `Provider`s](https://github.com/gradle/gradle/blob/1a078bd9b5cc4b04abab6fdfbed291cc890c8180/subprojects/core-api/src/main/java/org/gradle/api/Project.java#L956) support `Callable`s, and therefore Groovy closures as well. Lambdas are supported there as well, since `Callable` is a SAM:
```
project.provider(() -> 5)
```

The advantage of `Callable` is that it allows throwing a checked exception from the method.

`Provider` is related to the type (a delayed execution), though it still needs to be constructed. And I think this is where `Callable`, `Supplier`, `Closure` and the likes come into play. It looks to me that for construction of a `Provider` any SAM type is good, since it allow using a nice API from Groovy, Kotlin and Java. I think we could switch the supported SAM type from `Callable` to `Supplier`. Though I also think that if we switch the SAM type accepted by the Provider factory methods, then we should switch to a type owned by Gradle and not a JVM type.

In conclusion of all the above, why do you think that `Supplier` is better than `Callable` on the Gradle public API?

**Comment by lptr** [comment]
I think we are talking about different things here. What you are saying makes sense to me (i.e. don't bother replacing the current SAM types in APIs because that doesn't give us any advantage). What I'm saying is if we encounter a `Supplier` during the resolution of a `FileCollection` or a value, we should unpack it, just as we unpack `Callable`. So you could do these things:

```java

@Input
Supplier<String> getConfigFileName() { return () -> "config.txt"; }

project.files(getConfigFileName());
```

**Comment by wolfs** [comment]
Got it. I think we don't want to support `Supplier` as an `@Input` properties but use `Provider` instead. That is why I am reluctant to support it for resolving it in `FileCollection`s, since I don't see a good use case. For the use cases I came up with a `Provider` would be better.

**Comment by lptr** [comment]
Following that logic, we should also stop supporting `Callable` in favor of `Provider`, no?

**Comment by wolfs** [comment]
Yes, I think so.

**Comment by lptr** [comment]
I've updated the issue thus.
